File: Examples/Advanced/Archive-Job-Execution-Outputs/nodeJS/archive-job-execution-outputs.mjs
#!/usr/bin/env node

import fs from "node:fs";
import path from "node:path";
import process from "node:process";
import { randomUUID } from "node:crypto";
import { fileURLToPath, pathToFileURL } from "node:url";

export const STATE_FORMAT = "logicvein-netld-job-execution-output-state";
export const RUN_FORMAT = "logicvein-netld-job-execution-output-run";
export const FORMAT_VERSION = 1;

export class ExampleError extends Error {}

export function safeName(value, fallback) {
  return String(value).trim().replace(/[^A-Za-z0-9._-]+/g, "_").replace(/^[._]+|[._]+$/g, "") || fallback;
}

export function loadConfig(base = path.dirname(fileURLToPath(import.meta.url))) {
  const baseUrl = (process.env.NETLD_BASE_URL || "").trim();
  const apiKey = (process.env.NETLD_API_KEY || "").trim();
  if (!baseUrl || !apiKey) {
    throw new ExampleError("Set NETLD_BASE_URL and NETLD_API_KEY in the environment file.");
  }
  const pageSize = Number.parseInt(process.env.NETLD_PAGE_SIZE || "100", 10);
  if (!Number.isInteger(pageSize) || pageSize <= 0) {
    throw new ExampleError("NETLD_PAGE_SIZE must be a positive integer.");
  }
  const initialMode = (process.env.NETLD_INITIAL_MODE || "latest").trim().toLowerCase();
  if (!new Set(["latest", "all"]).has(initialMode)) {
    throw new ExampleError("NETLD_INITIAL_MODE must be latest or all.");
  }
  const destination = (name, fallback) => {
    const value = process.env[name] || fallback;
    return path.isAbsolute(value) ? value : path.join(base, value);
  };
  return {
    baseUrl: baseUrl.replace(/\/$/, ""),
    apiKey,
    outputDir: destination("NETLD_OUTPUT_DIR", "job-execution-outputs"),
    statePath: destination("NETLD_STATE_FILE", "job-execution-output-state.json"),
    reportPath: destination("NETLD_RUN_REPORT_FILE", "job-execution-output-run.json"),
    pageSize,
    initialMode,
    searchScheme: (process.env.NETLD_SEARCH_SCHEME || "").trim(),
    searchData: (process.env.NETLD_SEARCH_DATA || "").trim(),
    jobType: (process.env.NETLD_JOB_TYPE ?? "Script Tool Job").trim(),
    jobName: (process.env.NETLD_JOB_NAME || "").trim(),
  };
}

export class NetLDClient {
  constructor(config, fetchImpl = fetch, timeout = 30000) {
    this.config = config;
    this.fetch = fetchImpl;
    this.timeout = timeout;
    this.cookie = "";
  }

  async request(endpoint, options = {}) {
    const headers = {
      Authorization: `Bearer ${this.config.apiKey}`,
      ...(this.cookie ? { Cookie: this.cookie } : {}),
      ...(options.headers || {}),
    };
    let response;
    try {
      response = await this.fetch(`${this.config.baseUrl}${endpoint}`, {
        ...options,
        headers,
        redirect: "manual",
        signal: AbortSignal.timeout(this.timeout),
      });
    } catch {
      throw new ExampleError(`Could not reach ${this.config.baseUrl}.`);
    }
    const setCookie = response.headers.get("set-cookie");
    if (setCookie) this.cookie = setCookie.split(";", 1)[0];
    if (response.status >= 300 && response.status < 400) {
      throw new ExampleError(`Request redirected to ${response.headers.get("location") || ""}.`);
    }
    if (!response.ok) throw new ExampleError(`Request failed with HTTP ${response.status}.`);
    return response;
  }

  async login() {
    await this.request("/rest");
  }

  async call(method, params) {
    const response = await this.request("/rest", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ jsonrpc: "2.0", method, params, id: randomUUID() }),
    });
    let data;
    try {
      data = await response.json();
    } catch {
      throw new ExampleError(`${method} returned invalid JSON.`);
    }
    if (data.error) throw new ExampleError(`${method} failed: ${JSON.stringify(data.error)}`);
    return data.result;
  }

  async searchExecutionPage(offset, pageSize) {
    const result = await this.call("Scheduler.searchExecutions", {
      scheme: this.config.searchScheme,
      data: this.config.searchData,
      pageData: { offset, executionData: [], pageSize, total: 0 },
      sortColumn: "endTime",
      descending: true,
    });
    if (!result || !Array.isArray(result.executionData)) {
      throw new ExampleError("Scheduler.searchExecutions returned an invalid page.");
    }
    return result;
  }

  async executionDetails(executionId) {
    const result = await this.call("Plugins.getExecutionDetails", { executionId });
    if (result == null) return [];
    if (!Array.isArray(result)) {
      throw new ExampleError("Plugins.getExecutionDetails returned an invalid collection.");
    }
    return result;
  }

  async downloadDetail(executionId, recordId) {
    const query = new URLSearchParams({ executionId: String(executionId), recordId: String(recordId) });
    return Buffer.from(await (await this.request(`/servlet/pluginDetail?${query}`)).arrayBuffer());
  }
}

export function loadState(statePath) {
  if (!fs.existsSync(statePath)) return { format: STATE_FORMAT, formatVersion: FORMAT_VERSION };
  let state;
  try {
    state = JSON.parse(fs.readFileSync(statePath, "utf8"));
  } catch (error) {
    throw new ExampleError(`Could not read archive state: ${statePath}`);
  }
  if (state.format !== STATE_FORMAT || state.formatVersion !== FORMAT_VERSION) {
    throw new ExampleError("The job-execution state file has an unsupported format.");
  }
  return state;
}

export function writeAtomic(destination, content) {
  fs.mkdirSync(path.dirname(destination), { recursive: true });
  const temporary = `${destination}.${process.pid}.${randomUUID()}.tmp`;
  try {
    fs.writeFileSync(temporary, content);
    fs.renameSync(temporary, destination);
  } finally {
    if (fs.existsSync(temporary)) fs.unlinkSync(temporary);
  }
}

export function writeJson(destination, value) {
  writeAtomic(destination, JSON.stringify(value, null, 2) + "\n");
}

function executionEnd(execution) {
  return execution.endTime == null ? null : Number(execution.endTime);
}

function executionId(execution) {
  const value = execution.id ?? execution.executionId;
  if (value == null) throw new ExampleError("Scheduler.searchExecutions returned a record without an execution ID.");
  return Number(value);
}

function eligible(config, execution) {
  return (!config.jobType || execution.jobType === config.jobType) &&
    (!config.jobName || execution.jobName === config.jobName);
}

export async function collectExecutions(client, config, state) {
  const watermark = state.lastEndTime == null ? null : Number(state.lastEndTime);
  const watermarkIds = new Set((state.executionIdsAtLastEndTime || []).map(Number));
  const observed = [];
  let offset = 0;
  let reportedTotal = null;

  while (reportedTotal === null || offset < reportedTotal) {
    const page = await client.searchExecutionPage(offset, config.pageSize);
    const batch = page.executionData;
    const pageOffset = Number(page.offset ?? offset);
    const pageTotal = Number(page.total ?? pageOffset + batch.length);
    reportedTotal = reportedTotal === null ? pageTotal : Math.max(reportedTotal, pageTotal);
    let passedWatermark = false;
    for (const execution of batch) {
      const endTime = executionEnd(execution);
      if (endTime === null) continue;
      if (watermark !== null && endTime < watermark) {
        passedWatermark = true;
        break;
      }
      observed.push(execution);
    }
    if (passedWatermark || (watermark === null && config.initialMode === "latest")) break;
    const nextOffset = pageOffset + batch.length;
    if (nextOffset >= reportedTotal) break;
    if (!batch.length || nextOffset <= offset) {
      throw new ExampleError("Execution paging stopped before all results were returned.");
    }
    offset = nextOffset;
  }

  if (watermark === null && config.initialMode === "latest") return { observed, candidates: [] };
  const candidates = observed.filter((execution) => {
    const endTime = executionEnd(execution);
    const isNew = endTime !== null && (
      watermark === null || endTime > watermark ||
      (endTime === watermark && !watermarkIds.has(executionId(execution)))
    );
    return isNew && eligible(config, execution);
  }).sort((left, right) => executionEnd(left) - executionEnd(right) || executionId(left) - executionId(right));
  return { observed, candidates };
}

export async function archiveExecution(client, config, execution) {
  const id = executionId(execution);
  const endTime = executionEnd(execution);
  if (endTime === null) throw new ExampleError(`Execution ${id} has not completed.`);
  const date = new Date(endTime).toISOString().slice(0, 10);
  const directory = path.join(config.outputDir, date, `${id}_${safeName(execution.jobName || "job", "job")}`);
  const details = await client.executionDetails(id);
  const outputs = [];
  for (const detail of details) {
    const detailId = Number(detail.id);
    const identity = safeName(`${detail.managedNetwork || "network"}_${detail.ipAddress || "device"}`, "device");
    const contentPath = path.join(directory, `${detailId}_${identity}.log`);
    const metadataPath = path.join(directory, `${detailId}_${identity}.metadata.json`);
    const content = await client.downloadDetail(id, detailId);
    writeAtomic(contentPath, content);
    writeJson(metadataPath, { executionId: id, detail, contentFile: path.basename(contentPath) });
    outputs.push({
      detailId,
      bytes: content.length,
      contentFile: path.relative(config.outputDir, contentPath),
      metadataFile: path.relative(config.outputDir, metadataPath),
    });
  }
  const metadataPath = path.join(directory, "execution.metadata.json");
  writeJson(metadataPath, { execution, outputCount: outputs.length });
  return {
    executionId: id,
    endTime,
    jobName: execution.jobName,
    outputCount: outputs.length,
    metadataFile: path.relative(config.outputDir, metadataPath),
    outputs,
  };
}

function advanceState(state, observed) {
  const completed = observed.filter((execution) => executionEnd(execution) !== null);
  if (!completed.length) return;
  const newest = Math.max(...completed.map(executionEnd));
  const ids = new Set(completed.filter((execution) => executionEnd(execution) === newest).map(executionId));
  if (state.lastEndTime != null && Number(state.lastEndTime) === newest) {
    for (const id of state.executionIdsAtLastEndTime || []) ids.add(Number(id));
  }
  state.lastEndTime = newest;
  state.executionIdsAtLastEndTime = [...ids].sort((left, right) => left - right);
}

export async function archiveJobExecutionOutputs(client, config, timestamp = new Date().toISOString()) {
  const state = loadState(config.statePath);
  const initialBaseline = state.lastEndTime == null && config.initialMode === "latest";
  const { observed, candidates } = await collectExecutions(client, config, state);
  const archived = [];
  const failures = [];
  for (const execution of candidates) {
    try {
      archived.push(await archiveExecution(client, config, execution));
    } catch (error) {
      failures.push({ executionId: execution.id, error: error.message });
    }
  }
  if (!failures.length) advanceState(state, observed);
  const report = {
    format: RUN_FORMAT,
    formatVersion: FORMAT_VERSION,
    generatedAt: timestamp,
    initialBaseline,
    observedCount: observed.length,
    candidateCount: candidates.length,
    archivedCount: archived.length,
    outputCount: archived.reduce((sum, item) => sum + item.outputCount, 0),
    failureCount: failures.length,
    archived,
    failures,
  };
  writeJson(config.statePath, state);
  writeJson(config.reportPath, report);
  return report;
}

export async function main() {
  try {
    const config = loadConfig();
    const client = new NetLDClient(config);
    await client.login();
    const report = await archiveJobExecutionOutputs(client, config);
    if (report.initialBaseline) console.log("Recorded the latest completed execution as the initial baseline.");
    console.log(
      `Archived ${report.archivedCount} executions and ${report.outputCount} outputs; ` +
      `recorded ${report.failureCount} failures in ${config.reportPath}`,
    );
    return report.failureCount ? 2 : 0;
  } catch (error) {
    console.error(`Error: ${error.message}`);
    return 1;
  }
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  process.exitCode = await main();
}

File: Examples/Advanced/Archive-Job-Execution-Outputs/Python/archive_job_execution_outputs.py
#!/usr/bin/env python3
"""Incrementally archive output from completed netLD/ThirdEye job executions."""

from __future__ import annotations

import argparse
import json
import os
import re
import sys
import tempfile
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import requests
from dotenv import load_dotenv


STATE_FORMAT = "logicvein-netld-job-execution-output-state"
RUN_FORMAT = "logicvein-netld-job-execution-output-run"
FORMAT_VERSION = 1


class ExampleError(RuntimeError):
    pass


def safe_name(value: str, fallback: str) -> str:
    cleaned = re.sub(r"[^A-Za-z0-9._-]+", "_", value.strip()).strip("._")
    return cleaned or fallback


def generated_at() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


@dataclass(frozen=True)
class Config:
    base_url: str
    api_key: str
    output_dir: Path
    state_path: Path
    report_path: Path
    page_size: int
    initial_mode: str
    search_scheme: str
    search_data: str
    job_type: str
    job_name: str

    @classmethod
    def from_env(cls, env_path: Path, base: Path) -> "Config":
        load_dotenv(env_path, override=True)
        url = os.getenv("NETLD_BASE_URL", "").strip()
        key = os.getenv("NETLD_API_KEY", "").strip()
        if not url or not key:
            raise ExampleError("Set NETLD_BASE_URL and NETLD_API_KEY in the environment file.")

        def destination(name: str, default: str) -> Path:
            value = Path(os.getenv(name, default))
            return value if value.is_absolute() else base / value

        try:
            page_size = int(os.getenv("NETLD_PAGE_SIZE", "100"))
        except ValueError as exc:
            raise ExampleError("NETLD_PAGE_SIZE must be a positive integer.") from exc
        if page_size <= 0:
            raise ExampleError("NETLD_PAGE_SIZE must be a positive integer.")
        initial_mode = os.getenv("NETLD_INITIAL_MODE", "latest").strip().lower()
        if initial_mode not in {"latest", "all"}:
            raise ExampleError("NETLD_INITIAL_MODE must be latest or all.")
        return cls(
            url.rstrip("/"),
            key,
            destination("NETLD_OUTPUT_DIR", "job-execution-outputs"),
            destination("NETLD_STATE_FILE", "job-execution-output-state.json"),
            destination("NETLD_RUN_REPORT_FILE", "job-execution-output-run.json"),
            page_size,
            initial_mode,
            os.getenv("NETLD_SEARCH_SCHEME", "").strip(),
            os.getenv("NETLD_SEARCH_DATA", "").strip(),
            os.getenv("NETLD_JOB_TYPE", "Script Tool Job").strip(),
            os.getenv("NETLD_JOB_NAME", "").strip(),
        )


class NetLDClient:
    def __init__(self, config: Config, timeout: float = 30):
        self.config = config
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"Authorization": f"Bearer {config.api_key}"})

    def request(self, method: str, endpoint: str, **kwargs: Any):
        try:
            response = self.session.request(
                method,
                f"{self.config.base_url}{endpoint}",
                timeout=self.timeout,
                allow_redirects=False,
                **kwargs,
            )
        except requests.RequestException as exc:
            raise ExampleError(f"Could not reach {self.config.base_url}.") from exc
        if 300 <= response.status_code < 400:
            raise ExampleError(f"Request redirected to {response.headers.get('Location', '')}.")
        try:
            response.raise_for_status()
        except requests.RequestException as exc:
            raise ExampleError(f"Request failed with HTTP {response.status_code}.") from exc
        return response

    def login(self) -> None:
        self.request("GET", "/rest")

    def call(self, method: str, parameters: dict[str, Any]) -> Any:
        response = self.request(
            "POST",
            "/rest",
            headers={"Content-Type": "application/json"},
            json={"jsonrpc": "2.0", "method": method, "params": parameters, "id": str(uuid.uuid4())},
        )
        try:
            data = response.json()
        except ValueError as exc:
            raise ExampleError(f"{method} returned invalid JSON.") from exc
        if data.get("error"):
            raise ExampleError(f"{method} failed: {json.dumps(data['error'])}")
        return data.get("result")

    def search_execution_page(self, offset: int, page_size: int) -> dict[str, Any]:
        result = self.call(
            "Scheduler.searchExecutions",
            {
                "scheme": self.config.search_scheme,
                "data": self.config.search_data,
                "pageData": {"offset": offset, "executionData": [], "pageSize": page_size, "total": 0},
                "sortColumn": "endTime",
                "descending": True,
            },
        )
        if not isinstance(result, dict) or not isinstance(result.get("executionData"), list):
            raise ExampleError("Scheduler.searchExecutions returned an invalid page.")
        return result

    def execution_details(self, execution_id: int) -> list[dict[str, Any]]:
        result = self.call("Plugins.getExecutionDetails", {"executionId": execution_id})
        if result is None:
            return []
        if not isinstance(result, list):
            raise ExampleError("Plugins.getExecutionDetails returned an invalid collection.")
        return result

    def download_detail(self, execution_id: int, record_id: int) -> bytes:
        return self.request(
            "GET",
            "/servlet/pluginDetail",
            params={"executionId": execution_id, "recordId": record_id},
        ).content


def load_state(path: Path) -> dict[str, Any]:
    if not path.exists():
        return {"format": STATE_FORMAT, "formatVersion": FORMAT_VERSION}
    try:
        state = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError) as exc:
        raise ExampleError(f"Could not read archive state: {path}") from exc
    if state.get("format") != STATE_FORMAT or state.get("formatVersion") != FORMAT_VERSION:
        raise ExampleError("The job-execution state file has an unsupported format.")
    return state


def write_atomic(path: Path, content: bytes) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary: Path | None = None
    try:
        with tempfile.NamedTemporaryFile(
            dir=path.parent, prefix=f".{path.name}.", suffix=".tmp", delete=False
        ) as stream:
            temporary = Path(stream.name)
            stream.write(content)
        os.replace(temporary, path)
    except Exception:
        if temporary:
            temporary.unlink(missing_ok=True)
        raise


def write_json(path: Path, value: dict[str, Any]) -> None:
    write_atomic(path, (json.dumps(value, indent=2, sort_keys=True, ensure_ascii=False) + "\n").encode())


def execution_end(execution: dict[str, Any]) -> int | None:
    value = execution.get("endTime")
    return int(value) if value is not None else None


def execution_id(execution: dict[str, Any]) -> int:
    value = execution.get("id", execution.get("executionId"))
    if value is None:
        raise ExampleError("Scheduler.searchExecutions returned a record without an execution ID.")
    return int(value)


def eligible(config: Config, execution: dict[str, Any]) -> bool:
    return (not config.job_type or execution.get("jobType") == config.job_type) and (
        not config.job_name or execution.get("jobName") == config.job_name
    )


def is_new_execution(
    execution: dict[str, Any], watermark: int | None, watermark_ids: set[int]
) -> bool:
    end_time = execution_end(execution)
    if end_time is None:
        return False
    return (
        watermark is None
        or end_time > watermark
        or (end_time == watermark and execution_id(execution) not in watermark_ids)
    )


def collect_executions(
    client: Any, config: Config, state: dict[str, Any]
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    watermark = int(state["lastEndTime"]) if state.get("lastEndTime") is not None else None
    watermark_ids = {int(value) for value in state.get("executionIdsAtLastEndTime", [])}
    observed: list[dict[str, Any]] = []
    offset = 0
    reported_total: int | None = None

    while reported_total is None or offset < reported_total:
        page = client.search_execution_page(offset, config.page_size)
        batch = page["executionData"]
        page_offset = int(page.get("offset", offset))
        page_total = int(page.get("total", page_offset + len(batch)))
        reported_total = page_total if reported_total is None else max(reported_total, page_total)
        passed_watermark = False
        for execution in batch:
            end_time = execution_end(execution)
            if end_time is None:
                continue
            if watermark is not None and end_time < watermark:
                passed_watermark = True
                break
            observed.append(execution)

        if passed_watermark or (watermark is None and config.initial_mode == "latest"):
            break
        next_offset = page_offset + len(batch)
        if next_offset >= reported_total:
            break
        if not batch or next_offset <= offset:
            raise ExampleError("Execution paging stopped before all results were returned.")
        offset = next_offset

    if watermark is None and config.initial_mode == "latest":
        return observed, []

    candidates = [
        execution
        for execution in observed
        if is_new_execution(execution, watermark, watermark_ids)
        and eligible(config, execution)
    ]
    candidates.sort(key=lambda item: (execution_end(item) or 0, execution_id(item)))
    return observed, candidates


def archive_execution(client: Any, config: Config, execution: dict[str, Any]) -> dict[str, Any]:
    exec_id = execution_id(execution)
    end_time = execution_end(execution)
    if end_time is None:
        raise ExampleError(f"Execution {exec_id} has not completed.")
    date = datetime.fromtimestamp(end_time / 1000, timezone.utc).strftime("%Y-%m-%d")
    directory = config.output_dir / date / f"{exec_id}_{safe_name(str(execution.get('jobName', 'job')), 'job')}"
    details = client.execution_details(exec_id)
    outputs = []
    for detail in details:
        detail_id = int(detail["id"])
        identity = safe_name(
            f"{detail.get('managedNetwork', 'network')}_{detail.get('ipAddress', 'device')}",
            "device",
        )
        content_path = directory / f"{detail_id}_{identity}.log"
        metadata_path = directory / f"{detail_id}_{identity}.metadata.json"
        content = client.download_detail(exec_id, detail_id)
        write_atomic(content_path, content)
        write_json(metadata_path, {"executionId": exec_id, "detail": detail, "contentFile": content_path.name})
        outputs.append(
            {
                "detailId": detail_id,
                "bytes": len(content),
                "contentFile": str(content_path.relative_to(config.output_dir)),
                "metadataFile": str(metadata_path.relative_to(config.output_dir)),
            }
        )
    execution_path = directory / "execution.metadata.json"
    write_json(execution_path, {"execution": execution, "outputCount": len(outputs)})
    return {
        "executionId": exec_id,
        "endTime": end_time,
        "jobName": execution.get("jobName"),
        "outputCount": len(outputs),
        "metadataFile": str(execution_path.relative_to(config.output_dir)),
        "outputs": outputs,
    }


def advance_state(state: dict[str, Any], observed: list[dict[str, Any]]) -> None:
    completed = [execution for execution in observed if execution_end(execution) is not None]
    if not completed:
        return
    newest = max(execution_end(execution) or 0 for execution in completed)
    ids = {execution_id(execution) for execution in completed if execution_end(execution) == newest}
    if state.get("lastEndTime") is not None and newest == int(state["lastEndTime"]):
        ids.update(int(value) for value in state.get("executionIdsAtLastEndTime", []))
    state.update(lastEndTime=newest, executionIdsAtLastEndTime=sorted(ids))


def archive_job_execution_outputs(
    client: Any, config: Config, timestamp: str | None = None
) -> dict[str, Any]:
    state = load_state(config.state_path)
    initial_baseline = state.get("lastEndTime") is None and config.initial_mode == "latest"
    observed, candidates = collect_executions(client, config, state)
    archived, failures = [], []
    for execution in candidates:
        try:
            archived.append(archive_execution(client, config, execution))
        except Exception as exc:
            failures.append({"executionId": execution.get("id"), "error": str(exc)})
    if not failures:
        advance_state(state, observed)
    report = {
        "format": RUN_FORMAT,
        "formatVersion": FORMAT_VERSION,
        "generatedAt": timestamp or generated_at(),
        "initialBaseline": initial_baseline,
        "observedCount": len(observed),
        "candidateCount": len(candidates),
        "archivedCount": len(archived),
        "outputCount": sum(item["outputCount"] for item in archived),
        "failureCount": len(failures),
        "archived": archived,
        "failures": failures,
    }
    write_json(config.state_path, state)
    write_json(config.report_path, report)
    return report


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--env", type=Path, default=Path(__file__).with_name(".env"))
    args = parser.parse_args()
    try:
        base = Path(__file__).resolve().parent
        config = Config.from_env(args.env, base)
        client = NetLDClient(config)
        client.login()
        report = archive_job_execution_outputs(client, config)
        if report["initialBaseline"]:
            print("Recorded the latest completed execution as the initial baseline.")
        print(
            f"Archived {report['archivedCount']} executions and {report['outputCount']} outputs; "
            f"recorded {report['failureCount']} failures in {config.report_path}"
        )
        return 2 if report["failureCount"] else 0
    except (ExampleError, OSError, ValueError, KeyError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
